src/gleam/iterator.gleam

```diff
@@ -662,3 +667,66 @@ pub fn zip(left: Iterator(a), right: Iterator(b)) -> Iterator(tuple(a, b)) {
   do_zip(left.continuation, right.continuation)
   |> Iterator
 }
+
+type ChunkBy(element, key) {
+  AnotherBy(List(element), key, element, fn() -> Action(element))
+  LastBy(List(element))
+  NoneBy
+}
+
+fn next_chunk_by(
+  continuation: fn() -> Action(element),
+  f: fn(element) -> key,
+  previous_key: key,
+  current_chunk: List(element),
+) -> ChunkBy(element, key) {
+  case continuation() {
+    Stop ->
+      case current_chunk {
+        [] -> NoneBy
+        remaining -> LastBy(list.reverse(remaining))
+      }
+    Continue(e, next) -> {
+      let key = f(e)
+      case key == previous_key {
+        True -> next_chunk_by(next, f, key, [e, ..current_chunk])
+        False -> AnotherBy(list.reverse(current_chunk), key, e, next)
+      }
+    }
+  }
+}
+
+fn do_chunk_by(
+  continuation: fn() -> Action(element),
+  f: fn(element) -> key,
+  previous_key: key,
+  previous_element: element,
+) -> Action(List(element)) {
+  case next_chunk_by(continuation, f, previous_key, [previous_element]) {
+    NoneBy -> Stop
+    LastBy(chunk) -> Continue(chunk, stop)
+    AnotherBy(chunk, key, el, next) ->
+      Continue(chunk, fn() { do_chunk_by(next, f, key, el) })
+  }
+}
+
+/// Creates an iterator that emits chunks of elements
+/// for which `f` returns the same value.
+///
+/// ## Examples
+///
+///    > from_list([1, 2, 2, 3, 4, 4, 6, 7, 7]) |> chunk_by(fn(n) { n % 2 }) |> to_list
+///    [[1], [2, 2], [3], [4, 4, 6], [7, 7]]
+///
+pub fn chunk_by(
```

Comment:
I'm not sure about the name here. This doesn't read very well.

```rust
chunk_by(over: xs, with: f)
```

What might fit better?

Here's some initial thoughts from me:

```rust
key_chunk(over: xs, with: f)
logically_chunk(over: xs, with: f)
identity_chunk(over: xs, with: f)
```

Reply:
How about:

```rust
chunk_by(over: xs, key: f)
```
when labeled, it reads much more cleanly:

```rust
iter |> iterator.chunk_by(key: fn(x) { ... })
```

Reply:
It doesn't quite fit the Swift inheritied pattern of labels being prepositions rather than nouns.

I think my favourite is probably just `chunk(by: fn(x) { ... })`, though that does leave the question of what the fixed-size chunking function might be called.

```
xs |> sized_chunk(6)
xs |> chop(6)
xs |> divvy(6)
```

Reply:
`sized_chunk(over: xs, into: count)` seems to be the most descriptive